SeQuant/core/space.hpp

```diff
@@ -417,109 +314,446 @@ class IndexSpace {
   Type type() const noexcept { return attr().type(); }
   QuantumNumbers qns() const noexcept { return attr().qns(); }
 
-  /// @brief returns the base key for IndexSpace objects
-  /// @param space an IndexSpace object
-  /// @throw bad_key if this space has not been registered
-  static std::wstring base_key(const IndexSpace &space) {
-    return base_key(space.attr());
-  }
 
-  /// @brief returns the base key for IndexSpace objects of the given attribute
-  /// @param attr the space attribute
-  /// @throw bad_attr if \p attr has not been registered
-  static std::wstring base_key(Attr attr) {
-    assert(attr.is_valid());
-    if (attr == Attr::null()) return L"";
-    if (!instance_exists(attr)) throw bad_attr();
-    return attr2basekey_.find(attr)->second;
-  }
 
   /// Default ctor creates space with nonnull type and null quantum numbers
-  IndexSpace() noexcept : attr_(nonnulltype, nullqns) {}
+  IndexSpace() {
+    attr_ = {{0x7fffffff}, nullqns};
+    base_key_ =L"";
+    approximate_size_ = 0;
+  }
 
   IndexSpace(const IndexSpace &other) {
     if (!other.attr().is_valid())
       throw std::invalid_argument(
           "IndexSpace copy ctor received invalid argument");
-    attr_ = other.attr_;
+  attr_ = other.get_attr();
+  base_key_ = other.get_base_key();
+  approximate_size_ = other.get_approximate_size();
   }
   IndexSpace(IndexSpace &&other) {
     if (!other.attr().is_valid())
       throw std::invalid_argument(
           "IndexSpace move ctor received invalid argument");
-    attr_ = other.attr_;
+    attr_ = other.get_attr();
+base_key_ = other.get_base_key();
+approximate_size_ = other.get_approximate_size();
   }
   IndexSpace &operator=(const IndexSpace &other) {
     if (!other.attr().is_valid())
       throw std::invalid_argument(
           "IndexSpace copy assignment operator received invalid argument");
-    attr_ = other.attr_;
+    attr_ = other.get_attr();
+    base_key_ = other.get_base_key();
+    approximate_size_ = other.get_approximate_size();
     return *this;
   }
   IndexSpace &operator=(IndexSpace &&other) {
     if (!other.attr().is_valid())
       throw std::invalid_argument(
           "IndexSpace move assignment operator received invalid argument");
-    attr_ = other.attr_;
+    attr_ = other.get_attr();
+    base_key_ = other.get_base_key();
+    approximate_size_ = other.get_approximate_size();
     return *this;
   }
 
-  /// @brief constructs an instance of an IndexSpace object
-  IndexSpace(Type type, QuantumNumbers qns) noexcept : attr_(type, qns) {
-    assert(attr_.is_valid());
+
+  const Attr get_attr()const {
```

Comment:
```suggestion
  Attr get_attr()const {
```
`const` on non-reference/pointer return values are usually not desirable: see also https://stackoverflow.com/a/8716406